Add ClientConfiguration Push upon connect

ndzik opened the pull request:
Instead of an elaborate handshake protocol, we settled for a simple `ClientConfig` which is pushed upon establishing a connection with the client. This `ClientConfig` includes the `NetworkID` to which the Operator is connected and `Address` of the corresponding contract.

Commit:
[operator] Added ClientPush logic

Before clients had to know the address of the Erdstall-Contract as well
as chainID to connect with the operator and be able to join the Erdstall
System.
This change allows the operator to push a `ClientConfig` over the wire
containing ChainID/NetworkID as well as Contract-Address. This allows a
client to set up all web3 related connections upon connection to theh
operator.

Signed-off-by: Norbert Dzikowski <norbert@perun.network>

Commit:
[operator] Adapted operator logic and config to new behaviour

Signed-off-by: Norbert Dzikowski <norbert@perun.network>

Commit:
[eth] Add NetworkID to Ethclient

Due to the `ClientConfig` push and the requirement to access the current
`NetworkID` of the Node, the `Ethclient` needs to be extended to hold
this information for later querying.

Signed-off-by: Norbert Dzikowski <norbert@perun.network>

Commit:
[cmd/client] Adapt client to extended EthClient constructor

Signed-off-by: Norbert Dzikowski <norbert@perun.network>

Commit:
[*test] Adjust all required tests to new logic

Due to the changes to the RPCOperator and Ethclient, various tests had
to be adjusted.

Signed-off-by: Norbert Dzikowski <norbert@perun.network>

Commit:
[demo] Add key and certificate path to demo config

Signed-off-by: Norbert Dzikowski <norbert@perun.network>

ggwpez commented on demo/config.json:
```diff
@@ -8,6 +8,8 @@
 	"PowDepth": 0,
 	"RpcPort": 8401,
 	"RpcHost": "0.0.0.0",
+        "KeyFile": "/etc/letsencrypt/live/demo.erdstall.dev/privkey.pem",
```
I think it would be cleaner to create a certificate for the `operator` instead of copying it from the `demo` server.

sebastianst replied:
I actually think we shouldn't set TLS paths in the `demo/config` file, but document the json fields. This is a path not relevant for this public repo.

sebastianst committed:
operator,client: Pass OpServerConfig to NewRPC, hide ServeMux from constructors

sebastianst commented on operator/operator.go:
```diff
@@ -172,9 +172,16 @@ func (operator *Operator) Serve(port uint16) error {
 
 	operator.rpcOperator = NewRPCOperator(operator.enclave)
 
+	ctxReq, cancel := eth.ContextNodeReq()
+	defer cancel()
```
You could move this into the new `NetworkID` method and remove the context from its parameters.

RmbRT committed:
eth: Made NetworkID a function of Client.

Signed-off-by: Steffen Rattay <steffen@perun.network>

sebastianst commented on README.md:
```diff
@@ -12,6 +12,10 @@ folder. All TEE related code can be found in `tee`.
 
 ## Getting started
 
+If you wish to enable TLS support for the websocket connection, go to
+`demo/config.json`, and set the fields `KeyFile` to the TLS private key file
+path, and `CertFile` to the TLS certificate file path.
```
```suggestion
If you wish to enable TLS support for the websocket connection, set the fields `KeyFile` to the TLS private key file
path, and `CertFile` to the TLS certificate file path in your configuration json file.
```
This is a general documentation about the configuration file, not specific to file `demo/config.json`. Maybe also put it in a separate sub section further down `### Configuration file`. Also mention that the configuration file can be specified for the op with cmdline flag `-config config.json`.

RmbRT committed:
demo: Removed custom TLS file paths from config.

- Added notice to README on how to configure TLS support.

Signed-off-by: Steffen Rattay <steffen@perun.network>

ggwpez reviewed (changes requested):
The failing CI seems to be indeed a bug introduced in this MR.

ggwpez reviewed (dismissed):
I dont review it but just wanted to unblock the MR.

RmbRT committed:
operator: Allowed ethClient.NetworkID() to fail.

Signed-off-by: Steffen Rattay <steffen@perun.network>

RmbRT committed:
operator: Softened block processor restrictions.

Now no longer crashes if a younger block than the big bang block is
entered. Instead, just ignores it and logs a warning.

Signed-off-by: Steffen Rattay <steffen@perun.network>

RmbRT committed:
client: Increased operator start time.

Previously, the operator some times did not start fast enough.

Signed-off-by: Steffen Rattay <steffen@perun.network>

ggwpez commented on operator/rpc.go:
```diff
@@ -77,6 +137,10 @@ func (r *RPCServer) connectionHandler(out http.ResponseWriter, in *http.Request)
 
 	peer := &Peer{conn: conn, op: r.op}
+	if err := peer.sendJSON(r.server.clientConfig); err != nil {
+		r.Log().WithError(err).Error("Pushing ClientConfig:", err)
```
You are logging the error twice. Just looked at it because i have to rebase onto it.

ndzik replied:
Good point, we also did the same mistake five lines above :laughing:

Commit:
[operator] Fixed redundant error logging in rpc

Signed-off-by: Norbert Dzikowski <norbert@perun.network>
